init.go

```diff
@@ -2,72 +2,146 @@
+		return err
 	}
 
-	// Retrieve the loaded OpenSSL version and check if it is supported.
-	// Notice that major and minor could not match with the version parameter
-	// in case the name of the shared library file differs from the OpenSSL
-	// version it contains.
-	imajor := int(C.go_openssl_version_major(handle))
-	iminor := int(C.go_openssl_version_minor(handle))
-	ipatch := int(C.go_openssl_version_patch(handle))
-	if imajor < 0 || iminor < 0 || ipatch < 0 {
-		return 0, 0, 0, errors.New("openssl: can't retrieve OpenSSL version")
-	}
-	major, minor, patch = uint(imajor), uint(iminor), uint(ipatch)
-	var supported bool
-	if major == 1 {
-		supported = minor == 1
-	} else if major == 3 {
-		// OpenSSL guarantees API and ABI compatibility within the same major version since OpenSSL 3.
-		supported = true
-	}
-	if !supported {
-		return 0, 0, 0, errUnsupportedVersion()
+	loadOpenSSLFuncs(handle)
+
+	// Initialize OpenSSL.
+	go_openssl_OPENSSL_init()
+	if go_openssl_OPENSSL_init_crypto(
+		_OPENSSL_INIT_ADD_ALL_CIPHERS|
+			_OPENSSL_INIT_ADD_ALL_DIGESTS|
+			_OPENSSL_INIT_LOAD_CONFIG|
+			_OPENSSL_INIT_LOAD_CRYPTO_STRINGS,
+		nil) != 1 {
+		dlclose(handle)
+		return fail("init crypto")
 	}
+	osslHandle = handle
+	return nil
+}
 
-	// Load the OpenSSL functions.
-	// See shims.go for the complete list of supported functions.
+// loadFuncs loads and initialize the OpenSSL functions.
+// See shims.go for the complete list of supported functions.
+func loadOpenSSLFuncs(handle unsafe.Pointer) {
 	mkcgoLoad_(handle)
-	if major == 1 {
+	if vMajor == 1 {
 		mkcgoLoad_legacy_1(handle)
-		if patch == 1 {
+		if vPatch == 1 {
 			mkcgoLoad_111(handle)
 		}
 	} else {
 		mkcgoLoad_111(handle)
 		mkcgoLoad_3(handle)
 	}
+}
 
-	// Initialize OpenSSL.
-	go_openssl_OPENSSL_init()
-	if go_openssl_OPENSSL_init_crypto(
-		_OPENSSL_INIT_ADD_ALL_CIPHERS|
-			_OPENSSL_INIT_ADD_ALL_DIGESTS|
-			_OPENSSL_INIT_LOAD_CONFIG|
-			_OPENSSL_INIT_LOAD_CRYPTO_STRINGS,
-		nil) != 1 {
-		return 0, 0, 0, fail("openssl: init crypto")
+// initForCheckVersion loads and initialize only the
+// functions required in [CheckVersion].
+// It returns a function that must be called to release the resources.
+// The function leaves all the global variables in the same state as they were
+// before the call.
+func initForCheckVersion(file string) (func(), error) {
+	// This function can be called when the
+	prevMajor, prevMinor, prevPatch := vMajor, vMinor, vPatch
+	// Load the OpenSSL shared library using dlopen.
+	handle, err := openLibrary(file)
+	if err != nil {
+		vMajor, vMinor, vPatch = prevMajor, prevMinor, prevPatch
+		return nil, err
+	}
+
+	switch vMajor {
+	case 1:
+		mkcgoLoad_init_1(handle)
+	case 3:
+		mkcgoLoad_init_3(handle)
+	default:
+		// We shouldn't get here: openLibrary should have already returned an error.
+		panic(errUnsupportedVersion())
+	}
+	return func() {
+		dlclose(handle)
+		// Undo all the changes made in this function.
+		mkcgoUnload_version()
```

Comment:
Could be a little more futureproof/self-evident if `openLibrary` returned a cancel func that gets called here and explicitly ignored in `opensslInit`.

Otherwise, arguably the comment should be one line up if we're talking about "in this function [and functions it called]", or one line down if "in this function [directly]".